pkg/antctl/command_definition.go

```diff
@@ -0,0 +1,425 @@
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/handlers"
+)
+
+type formatterType string
+
+const (
+	jsonFormatter  formatterType = "json"
+	yamlFormatter  formatterType = "yaml"
+	tableFormatter formatterType = "table"
+)
+
+// commandGroup is used to group commands, it could be specified in commandDefinition.
+// The default commandGroup of a commandDefinition is `flat` which means the command
+// is a direct sub-command of the root command. For any other commandGroup, the
+// antctl framework will generate a same name sub-command of the root command for
+// each of them, any commands specified as one of these group will need to be invoked
+// as:
+//   antctl <commandGroup> <command>
+type commandGroup uint
+
+const (
+	flat commandGroup = iota
+	get
+)
+
+var groupCommands = map[commandGroup]*cobra.Command{
+	get: {
+		Use:   "get",
+		Short: "Get the status or resource of a topic",
+		Long:  "Get the status or resource of a topic",
+	},
+}
+
+const (
+	// tagKey is the tag name of the antctl specific annotation.
+	// For example:
+	// 	type FooResponse struct {
+	// 		Bar BarType `antctl:"key"`
+	// 	}
+	// If the field is annotated with antctl:"key", the framework assumes this field
+	// could be used to retrieve a unique Response, thus the framework will generate
+	// corresponding arg options to the cobra.Command.
+	tagKey = "antctl"
+	// tagOptionKeyArg is the option for antctl annotation. It tells the antctl
+	// the field is a primary key.
+	tagOptionKeyArg = "key"
+)
+
+// argOption describes one argument which can be used in a RequestOption.
+type argOption struct {
+	name      string
+	fieldName string
+	usage     string
+	key       bool
+}
+
+// commandDefinition defines options to create a cobra.Command for a antctl client.
+type commandDefinition struct {
+	// Cobra related
+	Use     string // The lower value of it will be used as the endpoint path, like: <API group>/<lower(Use)>.
+	Short   string
+	Long    string
+	Example string // It will be filled with generated examples if it is not provided.
+	// commandGroup represents the group of the command.
+	CommandGroup commandGroup
+	// Controller should be true if this command works for antrea-controller.
+	Controller bool
+	// Agent should be true if this command works for antrea-agent.
+	Agent bool
+	// SingleObject should be true if the handler always returns single object. The
+	// antctl assumes the response data as a slice of the objects by default.
+	SingleObject bool
+	// API is the endpoint for the command to retrieve data.
+	API string
+	// The handler factory of the command.
+	HandlerFactory handlers.Factory
+	// GroupVersion is the group version of the command handler, it should be set
+	// alongside with the HandlerFactory.
+	GroupVersion *schema.GroupVersion
+	// TransformedResponse is the final response struct of the command. If the
```

Comment:
I am not sure I understand this comment fully, but I think what you mean is:
> TransformedResponse is used to identify the struct type of the response received from the server (and that we need to output in the desired format). This type information is used to convert the response received from the server to the appropriate struct. If AddonTransform is set, TransformedResponse is not needed as AddonTransform is in-charge of converting the response to a struct.

If the description above is correct, have you considered using a `reflect.Type` here instead if an `interface{}`?

Reply:
Sorry I didn't clarify it. The `TransformedResponse` is always be used to tell the formatter how to output the response. The `AddonTransform` is used to parse the response from a handler but not is not for outputting.

The suggestion is good, I have updated the code accordingly.

Reply:
Thanks for clarifying. It seems that the original comment implied incorrectly that the `TransformedResponse` was sometimes not needed.